### PR TITLE
fix(inductor): use fillDMA for constant tensor and caching

### PR DESCRIPTION
**Pull request by thoangtrvn**
## What type of PR is this?

- [ ] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

## What this PR does:

This PR optimizes `spyre_constant_tensor` by adding a caching mechanism that eliminates redundant tensor creation for identical constant values. It also uses fillDMA to perform constant tensor creation.

### Memory Bounded by Unique Constants

 Cache size is bounded by the number of unique constants in a model:
 - Typical model: tens to hundreds of unique constants
 - Memory per constant: ~128 bytes (0-d scalar tensor)
 - Total for 200 constants: ~25 KB (negligible)

 This is **not unbounded growth** - it's bounded by model complexity.

## Which issue(s) this PR is related to:

Issue #3153 

## Special notes for your reviewer:

Comparison Summary:

  - Old implementation: 695.5 µs/op
  - New (after first run): 2.8 µs/op (246x speedup, 99.6% improvement!)

### CODE REVIEW

**Review by dgrove-oss — CHANGES_REQUESTED** [review]
Summary

Speeds up spyre_constant_tensor (695µs → 2.8µs per the PR's own benchmark) by (1) replacing torch.tensor(...).to(device) with an on-device torch.empty + fill_tensor DMA fill, and (2) adding a process-lifetime, unbounded, module-level cache keyed by (value, device, dtype) so repeated identical constants reuse the same device tensor.

Review

BLOCKERs

- [ ] _constant_tensor_cache is a plain module-level dict with no eviction, no size bound, and no clearing hook — it lives for the life of the process, independent of torch.compiler.reset() / Dynamo cache resets. (torch_spyre/_inductor/op_spec.py:201) Every existing test in test_dedup_constants.py etc. isolates via torch.compiler.reset() between cases; this cache silently survives that, so test-to-test isolation for anything touching constant values is now weaker than the rest of the suite assumes. In a long-running process (e.g. a serving loop with many distinct torch.compile calls, or a test suite that runs thousands of parameterized cases with varying constant values), this is unbounded device-memory growth with no way to reclaim it.
- [ ] The returned tensor is the same shared object handed back to every caller with a matching key, not a fresh tensor or a defensive .clone(). SpyreConstantFallback.should_allocate()==False and get_mutation_names()==[] (torch_spyre/_inductor/ir.py:374-378) currently make Inductor itself treat this buffer as immutable/non-poolable, so today's callers are safe — but the safety is enforced entirely by that IR-level contract, not by anything in spyre_constant_tensor itself. A future change to any pass that touches SpyreConstantFallback buffers (there are many consumers — dedup_constants.py, coarse_tile.py, padding.py, work_division.py) which relaxes that contract would silently start corrupting every other call site sharing the same cached tensor, with no assertion here to catch it. Worth at least a comment on spyre_constant_tensor noting the returned tensor must never be mutated in place, since the invariant it depends on lives in a different file.

SUGGESTIONs

- [ ] No new tests. Given this both changes the fill mechanism (DMA vs CPU-tensor-then-copy) and adds a new caching layer with real correctness preconditions (see BLOCKER above), it should have at least one test that: (a) two spyre_constant_tensor calls with the same (value, device, dtype) return the same object, (b) different dtype/device/value produce distinct tensors, (c) the produced value is numerically correct after the DMA fill (compare against the old torch.tensor(...).to(device) path or CPU reference).
- [ ] import torch_spyre inside the function (op_spec.py:210) is redundant — torch_spyre._C is already imported at module scope (op_spec.py:22), so torch_spyre._C.fill_tensor could be used directly without a per-call import.
- [ ] float(const_val) collapses NaN-valued constants to always-miss the cache (float('nan') != float('nan'), so no two NaN entries ever match) — harmless today (falls back to always creating a fresh tensor) but worth a one-line comment so a future reader doesn't mistake it for a bug.
- [ ] Comparison numbers in the PR description ("695.5 µs/op" old vs "2.8 µs/op (after first run)") conflate the DMA-fill speedup with the cache-hit speedup — it'd help reviewers to see the two effects separated (cache-miss-but-DMA-fill vs cache-hit) since the BLOCKER above is specifically about the caching half.

QUESTIONs

- [ ] Is this cache safe across muli-process or multi-threadedcompilation? str(device) is used as part of the key — is that guaranteed stable and unambiguous
(e.g. does torch.device("spyre") vstringify consistently with whateverthe wrapper codegen emits at wrapper.py:134)?
- [ ] Should this cache be scoped t lifetime (e.g. per-compiled-graph,or cleared via a hook into torch._dynamo.reset()/torch.compiler.reset()) rather than being
process-global forever? If the intempilation," a functools.lru_cachewith an explicit bound, or storage on V.graph, would give a natural scope and eviction policy for
free.

Verdict

REQUEST_CHANGES

The DMA-fill optimization itself losured win. The caching layer is theconcern: it's a process-lifetime, unbounded, un-cleared global cache backing a correctness     invariant (safe-to-share tensor) thR-level contract in a different file, with no test coverage added. Bounding/scoping the cache lifetime (or at minimum wiring it into the existing reset path) and addingests would resolve the blockers.

**Comment by thoangtrvn** [comment]
Thanks @dgrove-oss . Good points:
* I re-scoped the cache to the individual graph only 
* I also added a set of new tests to check the new behavior, i.e., a constant tensor is cached per graph and is freed when the graph is removed (a different graph won't see it), and the cache is based on value and dtype.

File: torch_spyre/_inductor/op_spec.py [hunk 1]
```diff
@@ -267,7 +268,54 @@ class LoopSpec:
...
+
+    Note:
+
+        float(const_val) makes NaN values always miss the cache because
+        float('nan') != float('nan'). This is intentional - it falls back to
+        creating a fresh tensor for each NaN request rather than attempting
+        cache matching.
+    """
+    # Normalize device to string
+    device_type = device.type if hasattr(device, "type") else str(device)
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/op_spec.py]: if device_type resolves to "spyre:0", this will fall to the cpu path

File: torch_spyre/_inductor/op_spec.py [hunk 2]
```diff
@@ -267,7 +268,54 @@ class LoopSpec:
...
+
+    # For non-Spyre devices (e.g., CPU), use standard torch.tensor path
+    if device_type != "spyre":
+        return torch.tensor(const_val, dtype=dtype).to(device)
+
+    from torch._inductor.virtualized import V
+
+    # Get or create per-graph cache (lives for the lifetime of the compilation)
+    cache = V.graph.__dict__.setdefault("_spyre_constant_tensors", {})
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/op_spec.py]: spyre_constant_tensor(...) is codegen'd into the wrapper and runs at model-execution time, where no Inductor GraphLowering is installed, so V.graph returns a fresh NullHandler() on every access. The cache dict is stashed on a throwaway object and discarded before the next call — every call is a miss (re-runs torch.empty + _C.fill_tensor).

Drop the runtime cache entirely; rely on the existing compile-time dedup_and_promote_constants pass (already wired in passes.py, and it has an output-aliasing guard). Keep only the torch.empty(()) + _C.fill_tensor change.

**Comment by tehbone** [comment]
Could this thing being fixed here be the reason why we're seeing so many 128-byte tensors being sent to the card? In a measured run (non-performance machine), this accounted for nearly 75% of the operation time.

**Comment by ani300** [comment]
@tehbone yes, it's not that bad on x86, but this will give a significant performance boost

File: torch_spyre/_inductor/op_spec.py (hunk 2 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/op_spec.py]: Thanks @JRosenkranz . This is a valid issue and is resolved now.

File: torch_spyre/_inductor/op_spec.py (hunk 1 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/op_spec.py]: Fixed this also.

File: torch_spyre/_inductor/op_spec.py [hunk 3]
```diff
@@ -239,6 +241,56 @@ class OpSpec:
...
 
 
+# --- Module-level constant tensor cache --------------------------------------
+#
+# Cache for Spyre constant tensors to avoid redundant tensor creation across
+# multiple forward passes.
+#
+# Thread-safe for concurrent inference workloads.
+#
+_CONSTANT_TENSOR_CACHE: dict[tuple, torch.Tensor] = {}
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/op_spec.py]: This is a strong-ref, unbounded, process-lifetime dict that never evicts and outlives torch.compiler.reset() — a long-running serving loop with many distinct constants leaks device memory with no way to reclaim it; make it a weakref.WeakValueDictionary (or a bounded LRU) so entries are freed when the graph drops them.

File: torch_spyre/_inductor/op_spec.py [hunk 4]
```diff
@@ -239,6 +241,56 @@ class OpSpec:
...
+
+    _original_reset = torch.compiler.reset
+
+    def _reset_with_cache_clear():
+        """Wrapped reset that clears Spyre constant cache."""
+        _original_reset()
+        clear_constant_tensor_cache()
+
+    # Apply monkey-patch
+    torch.compiler.reset = _reset_with_cache_clear
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/op_spec.py]: Importing this module permanently monkey-patches the public torch.compiler.reset process-wide (never restored) — invasive and surprising for anything else in the process — and it doesn't bound the cache's lifetime: production inference never calls reset(), so the cache still grows unbounded in serving. A self-managing WeakValueDictionary removes the need to patch a global API at all.

File: torch_spyre/_inductor/op_spec.py (hunk 1 above)
Comment (JRosenkranz) [reply torch_spyre/_inductor/op_spec.py]: I don't see the change, was this fixed?

**Comment by tehbone** [comment]
This might be a silly question, but why can't the node be updated to have the constant as separate node in the graph, and use that as a means of caching the content? Or does this pass happen after graph modification?

File: torch_spyre/_inductor/op_spec.py (hunk 1 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/op_spec.py]: I missed committing this code. It is in now.

File: torch_spyre/_inductor/op_spec.py (hunk 3 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/op_spec.py]: This scalar tensor is only created via torch.compile(), not a direct torch API used by the user. In practice, the number of scalar tensors won't get to an extreme number. Having said that, to address the concern, I now use an LRU cache to limit it to 1000 entries.

File: torch_spyre/_inductor/op_spec.py (hunk 4 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/op_spec.py]: No monkey patch needed after LRU cache is used.

**Comment by thoangtrvn** [comment]
> This might be a silly question, but why can't the node be updated to have the constant as separate node in the graph, and use that as a means of caching the content? Or does this pass happen after graph modification?

Not at all. That's actually a great question. I tried caching at the graph level in one of the commit in this PR and found that adding to graph level is ephemeral, and it is not retained at runtime as raised by @JRosenkranz. The module-level cache persists across compilation and runtime, and the LRU cache ensure the cache is scoped to the bound and cleanup as needed.

File: tests/inductor/test_constant_tensor_cache.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for spyre_constant_tensor module-level LRU caching.
+
+Tests verify:
+1. Cache hit: Same (value, device, dtype) returns same tensor object
+2. Cache miss: Different values/devices/dtypes create new tensors
+3. LRU eviction: Oldest entries evicted when cache exceeds max size
```
Comment (JRosenkranz) [inline tests/inductor/test_constant_tensor_cache.py]: is there a test for this?

Comment (thoangtrvn) [reply tests/inductor/test_constant_tensor_cache.py]: updated a test for cache eviction (bounded size cache)

**Comment by tehbone** [comment]
> > This might be a silly question, but why can't the node be updated to have the constant as separate node in the graph, and use that as a means of caching the content? Or does this pass happen after graph modification?
> 
> Not at all. That's actually a great question. I tried caching at the graph level in one of the commit in this PR and found that adding to graph level is ephemeral, and it is not retained at runtime as raised by @JRosenkranz. The module-level cache persists across compilation and runtime, and the LRU cache ensure the cache is scoped to the bound and cleanup as needed.

I guess I'm a little confused. I was thinking that the graph could be updated to be represented as a separate op to be optimized as opposed to providing data with the graph.

**Comment by thoangtrvn** [comment]
> > > This might be a silly question, but why can't the node be updated to have the constant as separate node in the graph, and use that as a means of caching the content? Or does this pass happen after graph modification?
> > 
> > 
> > Not at all. That's actually a great question. I tried caching at the graph level in one of the commit in this PR and found that adding to graph level is ephemeral, and it is not retained at runtime as raised by @JRosenkranz. The module-level cache persists across compilation and runtime, and the LRU cache ensure the cache is scoped to the bound and cleanup as needed.
> 
> I guess I'm a little confused. I was thinking that the graph could be updated to be represented as a separate op to be optimized as opposed to providing data with the graph.

I am not sure I follow the question, you don't want to recompile again the whole graph.

**Review by dgrove-oss — DISMISSED** [review]
All BLOCKERs and SUGGESTIONs from both reviewers appear resolved in the current diff.

Review

BLOCKERs

None.

SUGGESTIONs

- [ ] The LRU eviction and lookup/insert are each individually lock-protected (op_spec.py:340-356), but there's no test exercising eviction concurrently with lookups — test_thread_safety uses only 10 distinct values, well under _CACHE_MAX_SIZE, so it never touches the eviction path under contention. Not a correctness bug (the lock scope is correct — move_to_end/insert/evict all happen inside one with _CACHE_LOCK block per call), just a gap versus what the test name implies it covers.
- [ ] test_different_device_creates_new_tensor (test_constant_tensor_cache.py:102) asserts torch.device("spyre", 0) and torch.device("spyre") (bare, index=None) produce different cache entries. Per wrapper.py:132, node.layout.device is always a concretely-indexed device at the only real call site, so this distinction is untestable-in-practice and slightly misleading as documentation of intended behavior — consider a comment noting this is a defensive/theoretical case rather than one that occurs via codegen.
- [ ] _CACHE_MAX_SIZE = 1000  # Maximum number of cached constants (~128KB) — the size comment assumes fp16 scalars; worth noting it's an approximation (a torch.float64 scalar cache would be 4x that, still negligible either way, but the comment reads as precise when it isn't).

QUESTIONs

None — the two questions originally raised (multi-process/thread device-string stability, cache lifetime scoping) were both resolved by the switch to a real LRU with explicit bounds and clearing hooks, making them moot.

Verdict

APPROVE

Both dgrove-oss's and JRosenkranz's blocking concerns — unbounded process-lifetime growth, no test coverage, no clearing/isolation hook, the monkey-patch approach, and the redundant import — are resolved in the current diff via a genuine LRU cache with explicit _CACHE_MAX_SIZE, clear_constant_tensor_cache(), and new tests covering hit/miss/dtype/device/bounding/numerical-correctness/thread-safety. CI is green (including the new Test Constant Tensor Cache suite) and all 9 commits are signed off.

**Comment by dgrove-oss** [comment]
@thoangtrvn - need to resolve the merge conflict and then I think this is good to go.

**Review by ani300 — APPROVED** [review]
I've been testing the latest version of this code running granite in both FMS and hf-adapters and it's shown performance improvements and no broken tests.

**Review by dgrove-oss — APPROVED** [review]
lgtm
